planner/core/memtable_predicate_extractor.go

```diff
@@ -662,6 +676,113 @@ func (e *ClusterLogTableExtractor) explainInfo(p *PhysicalMemTable) string {
 	return s
 }
 
+// HotRegionsHistoryTableExtractor is used to extract some predicates of `tidb_hot_regions_history`
+type HotRegionsHistoryTableExtractor struct {
+	extractHelper
+
+	// SkipRequest means the where clause always false, we don't need to request any pd server.
+	SkipRequest bool
+
+	// StartTime represents the beginning time of update time.
+	// e.g: SELECT * FROM tidb_hot_regions_history WHERE update_time>'2019-10-10 10:10:10.999'
+	StartTime int64
+	// EndTime represents the ending time of update time.
+	// e.g: SELECT * FROM tidb_hot_regions_history WHERE update_time<'2019-10-11 10:10:10.999'
+	EndTime int64
+
+	// RegionIDs(StoreIDs) represents all region(store) ids we should filter in PD to reduce network IO.
+	// e.g:
+	// 1. SELECT * FROM tidb_hot_regions_history WHERE region_id=1
+	// 2. SELECT * FROM tidb_hot_regions_history WHERE table_id in (11, 22)
+	// Leave range operation to above selection executor.
+	RegionIDs []uint64
+	StoreIDs  []uint64
+	PeerIDs   []uint64
+	// Roles represents all roles (leader or follower) we should filter in PD to reduce network IO.
+	Roles []uint64
+
+	// HotRegionTypes represents all hot region types we should filter in PD to reduce network IO.
+	// e.g:
+	// 1. SELECT * FROM tidb_hot_regions_history WHERE type='read'
+	// 2. SELECT * FROM tidb_hot_regions_history WHERE type in ('read', 'write')
+	// 3. SELECT * FROM tidb_hot_regions_history WHERE type='read' and type='write' -> SkipRequest = true
+	HotRegionTypes set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *HotRegionsHistoryTableExtractor) Extract(
+	ctx sessionctx.Context,
+	schema *expression.Schema,
+	names []*types.FieldName,
+	predicates []expression.Expression,
+) []expression.Expression {
+	// Extract the `region_id/store_id/peer_id/type/roles` columns
+	remained, regionIDSkipRequest, regionIDs := e.extractCol(schema, names, predicates, "region_id", false)
+	remained, storeIDSkipRequest, storeIDs := e.extractCol(schema, names, remained, "store_id", false)
+	remained, peerIDSkipRequest, peerIDs := e.extractCol(schema, names, remained, "peer_id", false)
+	remained, isLeaderSkipRequest, roles := e.extractCol(schema, names, remained, "is_leader", false)
+	remained, typeSkipRequest, types := e.extractCol(schema, names, remained, "type", false)
+
+	e.SkipRequest = regionIDSkipRequest || storeIDSkipRequest || peerIDSkipRequest || typeSkipRequest || isLeaderSkipRequest
```

Comment:
If we get one skip, there is no need to extract the others.

Reply:
Is the existing fast return path enough?
```go
if e.SkipRequest {
      return nil
}
```

Reply:
how about:
```
remained, isLeaderSkipRequest, roles := e.extractCol(schema, names, remained, "is_leader", false)
if isLeaderSkipRequest {
    return nil
}
remained, typeSkipRequest, types := e.extractCol(schema, names, remained, "type", false)
if typeSkipRequest {
    return nil
}

```

Reply:
I got it.

Reply:
I check other extractor in this file, such as: `ClusterLogTableExtractor`, `InspectionSummaryTableExtractor` and `InspectionResultTableExtractor`. They use `||` operator, I think it is better to keep style with them. I will spilt them into smaller `||` grounp by field type.